### PR TITLE
Add abillity to get current merkle proof of an element

### PR DESCRIPTION
**Pull request by chris124567**
Added code along the lines of https://research.swtch.com/tlog so we could get the current merkle proof of a given element.

### CODE REVIEW

**Comment by chris124567** [comment]
Should be good to go.

File: internal/explorerutil/hashstore.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+func (h *HashStore) MerkleProof(leafIndex uint64) ([]types.Hash256, error) {
+	var indexBits [64]uint8
+	for i := 0; i < len(indexBits); i++ {
+		indexBits[i] = uint8((leafIndex & (1 << i)) >> i)
+	}
+
+	proof := make([]types.Hash256, trailingOnes(leafIndex))
+	for i := uint64(0); i < uint64(len(proof)); i++ {
+		if indexBits[i] == 0 {
+			leafIndex += pow2(i)
```
Comment (lukechampine) [inline internal/explorerutil/hashstore.go:37]: I would define `n := 1<<i` at the top of the loop and ditch the `pow2` function

File: internal/explorerutil/hashstore.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+// MerkleProof implements explorer.HashStore.
+func (h *HashStore) MerkleProof(leafIndex uint64) ([]types.Hash256, error) {
+	var indexBits [64]uint8
+	for i := 0; i < len(indexBits); i++ {
+		indexBits[i] = uint8((leafIndex & (1 << i)) >> i)
+	}
+
+	proof := make([]types.Hash256, trailingOnes(leafIndex))
+	for i := uint64(0); i < uint64(len(proof)); i++ {
+		if indexBits[i] == 0 {
```
Comment (lukechampine) [inline internal/explorerutil/hashstore.go:36]: instead do:
```go
if leafIndex&(1<<i) == 0
```
and define a separate `pos := leafIndex` that you can mutate

File: internal/explorerutil/hashstore.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			leafIndex -= pow2(i)
+		}
+		if _, h.err = h.hashFiles[i].ReadAt(proof[i][:], int64(leafIndex/(pow2(i)))*hashSize); h.err != nil {
+			return nil, h.err
+		}
+	}
+	return proof, nil
+}
+
+func (h *HashStore) ModifyLeaf(leafIndex uint64, proof []types.Hash256) error {
```
Comment (lukechampine) [inline internal/explorerutil/hashstore.go:48]: this can take a `types.StateElement`

File: internal/explorerutil/hashstore.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		}
+	}
+	return nil
+}
+
+// NewHashStore returns a new HashStore.
+func NewHashStore(dir string) (*HashStore, error) {
+	var h HashStore
+	for i := 0; i < 64; i++ {
+		hashFile, err := os.OpenFile(filepath.Join(dir, strconv.Itoa(i)), os.O_CREATE|os.O_RDWR, 0666)
```
Comment (lukechampine) [inline internal/explorerutil/hashstore.go:84]: let's use a more descriptive name: `fmt.Sprintf("tree_level_%d.dat")`

**Review by lukechampine — COMMENTED** [review]
great! Now we can set `MerkleProof = nil` on each element before storing it. (This might break your test, though.)

**Comment by chris124567** [comment]
:+1:

**Pull request merged**